### PR TITLE
Fix the check on isStoreIndirect in LocalAnalysis

### PR DESCRIPTION
**Pull request by a7ehuo**
In `hasOldExpressionOnRhs()`, we temporarily change wrtbar to aloadi
for the syntactic comparison in `areSyntacticallyEquivalent()`.
When it’s an indirect store, the number of children of the node is
set to 1, otherwise 0. Because the check on whether not the node
is an indirect store happens after the wrtbar node is recreated as aloadi,
`node->getOpCode().isStoreIndirect()` is false for aloadi, and the
number of children for wrtbar node ends up as 0. It prevents
`areSyntacticallyEquivalent()` from comparing the first child of the two
nodes. It concludes that different expressions as the same.
The fix is to check if it’s an indirect store from the original node.

Fixes eclipse/openj9#11569

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by fjeremic** [comment]
> In hasOldExpressionOnRhs(), we temporarily change wrtbar to aloadi
for the syntactic comparison in areSyntacticallyEquivalent().

This seems quite hacky and prone to errors as we're seeing in the defect that was opened. Is there a better way of doing this?

**Comment by a7ehuo** [comment]
>This seems quite hacky and prone to errors as we're seeing in the defect that was opened. Is there a better way of doing this?

My understanding is that it tries to treat loads and stores of the same field as the same expression[1]. @vijaysun-omr @liqunl  I wonder why `areSyntacticallyEquivalent()` wasn't initially implemented to handle such case and instead `hasOldExpressionOnRhs()` chose path to recreate/restore the original node? Would it make sense to update `areSyntacticallyEquivalent()` to treat loads/stores of the same field as the same? Or maybe create a new comparison API to handle the comparison between loads and stores to avoid recreating the node? Any potential issues? 

[1]
https://github.com/eclipse/omr/blob/bded46caf73287830139853961ffacb96cc9758e/compiler/optimizer/LocalAnalysis.cpp#L514-L519

**Comment by vijaysun-omr** [comment]
This is something that can be changed (i.e. it would be a somewhat localized change) but in my opinion, it is not worth the development effort. My reasoning is that there are hardly ever problems in this code in the many years that it has existed and the risk of introducing some other bug outweighs the benefit in my opinion. 

The earlier change that caused the issue that this PR is fixing ought to have checked if it was moving the changing of the opcode past a point that was checking the opcode in my opinion; changing the opcode is a valid operation after all (in a general sense) and so the fix that we have in this PR is the right one in my opinion.

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
@fjeremic any further comments from you since you had looked at this fix previously ? I am checking before merging since it has passed testing and is no longer WIP.

File: compiler/optimizer/LocalAnalysis.cpp [hunk 1]
```diff
@@ -488,7 +488,7 @@ int TR_LocalAnalysisInfo::hasOldExpressionOnRhs(TR::Node *node, bool recalcConta
 #endif
    bool seenWriteBarrier = false;
    int32_t storeNumChildren = node->getNumChildren();
-
+   bool isStoreIndirect = node->getOpCode().isStoreIndirect();
```
Comment (fjeremic) [inline compiler/optimizer/LocalAnalysis.cpp:491]: I'd like to see a comment here explaining why this value needs to be cached. We should explicitly state in the comment that the node will be recreated below and the question of whether the original store is indirect or not may change. We should also note the fact that the recreation is done for convenience of syntactic comparison.

This is non-trivial stuff happening here for future readers of the code, and we don't want them confused or folding away this variable should they not be as keen to the details here.

File: compiler/optimizer/LocalAnalysis.cpp [hunk 2]
```diff
@@ -548,6 +548,7 @@ int TR_LocalAnalysisInfo::hasOldExpressionOnRhs(TR::Node *node, bool recalcConta
       int32_t hashValue = _hashTable->hash(node);
       HashTable::Cursor cursor(_hashTable, hashValue);
       TR::Node *other;
+      bool tmpIsStoreIndirect = false;
```
Comment (fjeremic) [inline compiler/optimizer/LocalAnalysis.cpp:551]: I think we should rename this to something more meaningful, such as `isOtherStoreIndirect` and put a comment to reference the comment which you will make for `isStoreIndirect` as this follows the same pattern.

**Review by fjeremic — CHANGES_REQUESTED** [review]

**Comment by fjeremic** [comment]
> This is something that can be changed (i.e. it would be a somewhat localized change) but in my opinion, it is not worth the development effort. My reasoning is that there are hardly ever problems in this code in the many years that it has existed and the risk of introducing some other bug outweighs the benefit in my opinion.

I agree w.r.t. development effort. I want to explicitly call out that using the `TR::Node::recreate` API has the side-effect of potentially clearing some node properties:
https://github.com/eclipse/omr/blob/f38780c46dcc3c8d4725ee50f5cf23b5723ea72c/compiler/il/OMRNode.cpp#L527-L537

Some of this code is not active right now as it requires additional work, but I wanted to call out that changing an opcode of a node and changing it back is not a NOP operation as the properties may change.

**Comment by vijaysun-omr** [comment]
Okay I won't disagree with your notes and ultimately if someone did want to clean this up, I won't stand in the way by any means.  I was only giving my opinion on the risk-reward. 

I do agree with the non-trivial risk that some node flag issue does occur with the only mild reassurance being that we have not experienced such an issue in the twenty or so years that this code has been in existence.

File: compiler/optimizer/LocalAnalysis.cpp (hunk 1 above)
Comment (a7ehuo) [reply compiler/optimizer/LocalAnalysis.cpp:491]: Good point! I'll make the change.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by a7ehuo** [comment]
@fjeremic Comments addressed in 18b7e5d. Ready for another review. Thanks!

@vijaysun-omr `linux_390-64` build failed. It looks like related to infrastructure.

```
linux_390-64

Setting status of 18b7e5d9e46844dbecbaba7be400ac69774f4761 to FAILURE with url https://ci.eclipse.org/omr/job/PullRequest-linux_390-64/3076/ and message: 'Build finished. '
Using context: continuous-integration/eclipse-omr/pr/linux_390-64
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to ub16-s390x-01
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1788)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:998)
		at hudson.FilePath.act(FilePath.java:1069)
		at hudson.FilePath.act(FilePath.java:1058)
		at hudson.FilePath.mkdirs(FilePath.java:1243)
		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:79)
		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
java.nio.file.FileSystemException: /home/jenkins/workspace/Build: Read-only file system
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
	at java.nio.file.Files.createDirectory(Files.java:674)
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
	at java.nio.file.Files.createDirectories(Files.java:767)
	at hudson.FilePath.mkdirs(FilePath.java:3256)
	at hudson.FilePath.access$1300(FilePath.java:211)
	at hudson.FilePath$Mkdirs.invoke(FilePath.java:1251)
	at hudson.FilePath$Mkdirs.invoke(FilePath.java:1247)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3069)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:369)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

**Review by fjeremic — APPROVED** [review]

**Pull request merged**